### ISSUE TITLE
[Bug] Firmware updates not reflected in HA sensor for X1 printers

### ISSUE THREAD

**Issue by fmeus** [greghesp/ha-bambulab#227]
### Describe the bug

When the printer indicates that there is a firmware update available the sensor 'Firmware update' still hase the state 'Up-to-date'

### To Reproduce

Wait for a firmware update to be available for the printer and check the state of the binary_sensor.x1c_[sn]_firmware_update_available

to be sure either power cycle the printer or force refresh of data

### Expected Behaviour

Sensor to correctly indicate if there are firmware updates available

### What device are you using?

X1C

### Home Assistant Version

2023.9.2

### Component Version

v1.6.0-dev1

### Log Extracts

Diagnostic data at the time a firmware update was available (01.06.05.00 to 01.06.05.01)

```
{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.9.2",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.11.5",
    "docker": true,
    "arch": "aarch64",
    "timezone": "Europe/Amsterdam",
    "os_name": "Linux",
    "os_version": "6.1.45",
    "supervisor": "2023.09.2",
    "host_os": "Home Assistant OS 10.5",
    "docker_version": "23.0.6",
    "chassis": "embedded",
    "run_as_root": true
  },
  "custom_components": {
    "hacs": {
      "version": "1.32.1",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    },
    "sonoff": {
      "version": "3.5.2",
      "requirements": [
        "pycryptodome>=3.6.6"
      ]
    },
    "watchman": {
      "version": "0.5.1",
      "requirements": [
        "prettytable==3.0.0"
      ]
    },
    "matter_experimental": {
      "version": "0.3.0",
      "requirements": [
        "python-matter-server==0.3.0"
      ]
    },
    "webrtc": {
      "version": "v3.3.0",
      "requirements": []
    },
    "lghorizon": {
      "version": "0.5.0",
      "requirements": [
        "lghorizon>=0.6.2"
      ]
    },
    "smartthinq_sensors": {
      "version": "0.33.3",
      "requirements": [
        "pycountry>=20.7.3",
        "xmltodict>=0.12.0",
        "charset_normalizer>=2.0.0"
      ]
    },
    "anniversaries": {
      "version": "5.2.0",
      "requirements": [
        "python-dateutil>=2.8.1",
        "integrationhelper>=0.2.2",
        "voluptuous>=0.12.1"
      ]
    },
    "bambu_lab": {
      "version": "1.6.0",
      "requirements": []
    }
  },
  "integration_manifest": {
    "domain": "bambu_lab",
    "name": "Bambu Lab",
    "codeowners": [
      "@greghesp",
      "@AdrianGarside"
    ],
    "config_flow": true,
    "dependencies": [
      "mqtt",
      "ffmpeg"
    ],
    "documentation": "https://github.com/greghesp/ha-bambulab",
    "iot_class": "local_push",
    "issue_tracker": "https://github.com/greghesp/ha-bambulab/issues",
    "ssdp": [
      {
        "st": "urn:bambulab-com:device:3dprinter:1"
      }
    ],
    "version": "1.6.0",
    "is_built_in": false
  },
  "data": {
    "push_all": {
      "ams": {
        "ams": [
          {
            "humidity": "4",
            "id": "0",
            "temp": "34.5",
            "tray": [
              {
                "bed_temp": "0",
                "bed_temp_type": "0",
                "cali_idx": -1,
                "cols": [
                  "FFFFFFFF"
                ],
                "ctype": 0,
                "drying_temp": "0",
                "drying_time": "0",
                "id": "0",
                "nozzle_temp_max": "240",
                "nozzle_temp_min": "190",
                "remain": -1,
                "tag_uid": "0000000000000000",
                "tray_color": "FFFFFFFF",
                "tray_diameter": "0.00",
                "tray_id_name": "",
                "tray_info_idx": "GFL99",
                "tray_sub_brands": "",
                "tray_type": "PLA",
                "tray_uuid": "00000000000000000000000000000000",
                "tray_weight": "0",
                "xcam_info": "000000000000000000000000"
              },
              {
                "bed_temp": "45",
                "bed_temp_type": "1",
                "cali_idx": -1,
                "cols": [
                  "C00D1EFF"
                ],
                "ctype": 0,
                "drying_temp": "55",
                "drying_time": "8",
                "id": "1",
                "nozzle_temp_max": "230",
                "nozzle_temp_min": "190",
                "remain": 100,
                "tag_uid": "A012D35D00000100",
                "tray_color": "C00D1EFF",
                "tray_diameter": "1.75",
                "tray_id_name": "A00-R0",
                "tray_info_idx": "GFA00",
                "tray_sub_brands": "PLA Basic",
                "tray_type": "PLA",
                "tray_uuid": "371E7185CFFE498898E4AAAA660E8680",
                "tray_weight": "1000",
                "xcam_info": "D007D007E803E8036666663F"
              },
              {
                "bed_temp": "0",
                "bed_temp_type": "0",
                "cali_idx": -1,
                "cols": [
                  "161616FF"
                ],
                "ctype": 0,
                "drying_temp": "0",
                "drying_time": "0",
                "id": "2",
                "nozzle_temp_max": "270",
                "nozzle_temp_min": "220",
                "remain": -1,
                "tag_uid": "0000000000000000",
                "tray_color": "161616FF",
                "tray_diameter": "0.00",
                "tray_id_name": "",
                "tray_info_idx": "GFG99",
                "tray_sub_brands": "",
                "tray_type": "PETG",
                "tray_uuid": "00000000000000000000000000000000",
                "tray_weight": "0",
                "xcam_info": "000000000000000000000000"
              },
              {
                "bed_temp": "0",
                "bed_temp_type": "0",
                "cali_idx": -1,
                "cols": [
                  "DCDCDCFF"
                ],
                "ctype": 0,
                "drying_temp": "0",
                "drying_time": "0",
                "id": "3",
                "nozzle_temp_max": "270",
                "nozzle_temp_min": "220",
                "remain": -1,
                "tag_uid": "0000000000000000",
                "tray_color": "DCDCDCFF",
                "tray_diameter": "0.00",
                "tray_id_name": "",
                "tray_info_idx": "GFG99",
                "tray_sub_brands": "",
                "tray_type": "PETG",
                "tray_uuid": "00000000000000000000000000000000",
                "tray_weight": "0",
                "xcam_info": "000000000000000000000000"
              }
            ]
          }
        ],
        "ams_exist_bits": "1",
        "insert_flag": true,
        "power_on_flag": false,
        "tray_exist_bits": "f",
        "tray_is_bbl_bits": "f",
        "tray_now": "255",
        "tray_pre": "255",
        "tray_read_done_bits": "f",
        "tray_reading_bits": "0",
        "tray_tar": "255",
        "version": 13
      },
      "ams_rfid_status": 0,
      "ams_status": 0,
      "aux_part_fan": true,
      "bed_target_temper": 0.0,
      "bed_temper": 33.0,
      "big_fan1_speed": "0",
      "big_fan2_speed": "0",
      "cali_version": 0,
      "chamber_temper": 32.0,
      "command": "push_status",
      "cooling_fan_speed": "0",
      "fail_reason": "0",
      "fan_gear": 0,
      "filam_bak": [],
      "force_upgrade": false,
      "gcode_file": "",
      "gcode_file_prepare_percent": "0",
      "gcode_start_time": "0",
      "gcode_state": "IDLE",
      "heatbreak_fan_speed": "0",
      "hms": [],
      "home_flag": 117128,
      "hw_switch_state": 0,
      "ipcam": {
        "ipcam_dev": "1",
        "ipcam_record": "enable",
        "mode_bits": 2,
        "resolution": "1080p",
        "rtsp_url": "**REDACTED**",
        "timelapse": "disable",
        "tutk_server": "disable"
      },
      "layer_num": 0,
      "lifecycle": "product",
      "lights_report": [
        {
          "mode": "on",
          "node": "chamber_light"
        },
        {
          "mode": "flashing",
          "node": "work_light"
        }
      ],
      "maintain": 3,
      "mc_percent": 0,
      "mc_print_error_code": "0",
      "mc_print_stage": "1",
      "mc_print_sub_stage": 0,
      "mc_remaining_time": 0,
      "mess_production_state": "active",
      "msg": 0,
      "nozzle_diameter": "0.4",
      "nozzle_target_temper": 0.0,
      "nozzle_temper": 38.0,
      "nozzle_type": "hardened_steel",
      "online": {
        "ahb": false,
        "ext": false,
        "version": 9
      },
      "print_error": 0,
      "print_gcode_action": 0,
      "print_real_action": 0,
      "print_type": "",
      "profile_id": "",
      "project_id": "",
      "queue_est": 0,
      "queue_number": 0,
      "queue_sts": 0,
      "queue_total": 0,
      "s_obj": [],
      "sdcard": true,
      "sequence_id": "11",
      "spd_lvl": 2,
      "spd_mag": 100,
      "stg": [],
      "stg_cur": -1,
      "subtask_id": "",
      "subtask_name": "",
      "task_id": "",
      "total_layer_num": 0,
      "upgrade_state": {
        "ahb_new_version_number": "",
        "ams_new_version_number": "",
        "consistency_request": false,
        "dis_state": 1,
        "err_code": 0,
        "ext_new_version_number": "",
        "force_upgrade": false,
        "idx": 9,
        "message": "",
        "module": "null",
        "new_version_state": 1,
        "ota_new_version_number": "01.06.05.01",
        "progress": "0",
        "sequence_id": 0,
        "sn": "**REDACTED**",
        "status": "IDLE"
      },
      "upload": {
        "file_size": 0,
        "finish_size": 0,
        "message": "Good",
        "oss_url": "",
        "progress": 0,
        "sequence_id": "0903",
        "speed": 0,
        "status": "idle",
        "task_id": "",
        "time_remaining": 0,
        "trouble_id": ""
      },
      "vt_tray": {
        "bed_temp": "0",
        "bed_temp_type": "0",
        "cali_idx": -1,
        "cols": [
          "DCDCDCFF"
        ],
        "ctype": 0,
        "drying_temp": "0",
        "drying_time": "0",
        "id": "254",
        "nozzle_temp_max": "240",
        "nozzle_temp_min": "190",
        "remain": 0,
        "tag_uid": "0000000000000000",
        "tray_color": "DCDCDCFF",
        "tray_diameter": "0.00",
        "tray_id_name": "",
        "tray_info_idx": "GFL99",
        "tray_sub_brands": "",
        "tray_type": "PLA",
        "tray_uuid": "00000000000000000000000000000000",
        "tray_weight": "0",
        "xcam_info": "000000000000000000000000"
      },
      "wifi_signal": "-54dBm",
      "xcam": {
        "allow_skip_parts": false,
        "buildplate_marker_detector": true,
        "first_layer_inspector": true,
        "halt_print_sensitivity": "medium",
        "print_halt": true,
        "printing_monitor": true,
        "spaghetti_detector": true
      },
      "xcam_status": "0"
    },
    "get_version": {
      "command": "get_version",
      "module": [
        {
          "hw_ver": "",
          "name": "ota",
          "sn": "",
          "sw_ver": "01.06.05.00"
        },
        {
          "hw_ver": "AP05",
          "name": "rv1126",
          "sn": "**REDACTED**",
          "sw_ver": "00.00.24.12"
        },
        {
          "hw_ver": "TH09",
          "name": "th",
          "sn": "**REDACTED**",
          "sw_ver": "00.00.05.83"
        },
        {
          "hw_ver": "MC07",
          "name": "mc",
          "sn": "**REDACTED**",
          "sw_ver": "00.00.16.68/00.00.16.68"
        },
        {
          "hw_ver": "",
          "name": "xm",
          "sn": "",
          "sw_ver": "00.01.02.02"
        },
        {
          "hw_ver": "AMS08",
          "name": "ams/0",
          "sn": "**REDACTED**",
          "sw_ver": "00.00.06.40"
        }
      ],
      "sequence_id": "0"
    }
  }
}
```

**Comment by AdrianGarside** [comment]
We're looking for:
        self.firmware_updates = data.get("upgrade_state", {}).get("new_ver_list", self.firmware_updates)

And there is no new_ver_list in the mqtt any more it seems. Ah except I still have that on my P1P so maybe this only changed on the X1?
      "upgrade_state": {
        "sequence_id": 0,
        "progress": "",
        "status": "",
        "consistency_request": false,
        "dis_state": 0,
        "err_code": 0,
        "force_upgrade": false,
        "message": "",
        "module": "",
        "new_version_state": 2,
        "new_ver_list": []
      },

**Comment by fmeus** [comment]
Within the json data upgrade_state now has an attribute new_version_state. This appears to have a value of 2 when there are no updates and a value of 1 when there are updates available.

So far I have only been able to verify this for the firmware of the X1C, at this moment I have no information on the behaviour of new_version_state when there are multiple firmware updates available (i.e. printer + ams or multiple ams's).

The existing code was also checking for new_ver_list, this does not seem to appear anymore in the json data.

But made it is different per model, or has changed as of a certain firmware release?


I did a bit of testing with the following changes which seem to work fine for me. But given your comment for your P1P might have to start doing different check per model (or even firmware release)...

**pybambu/models.py**

    change datatype for firmware_updates from
        dict
      to
        int

    change default/init value for firmware_updates from
        {}
      to
        2 (no updates available)

    change assignment self.firmware_updates from 
        self.firmware_updates = data.get("upgrade_state", {}).get("new_ver_list", self.firmware_updates)
      to
        self.firmware_updates = data.get("upgrade_state", {}).get("new_version_state", self.firmware_updates)

**definitions.py**
    change to assignment for is_on_fn (in BambuLabBinarySensorEntityDescription for firmware_update) from 
        is_on_fn=lambda self: len(self.coordinator.get_model().info.firmware_updates) != 0
      to
        is_on_fn=lambda self: (self.coordinator.get_model().info.firmware_updates) != 2

**Comment by AdrianGarside** [comment]
The only X1 mqtt payload I have to hand is probably pretty old at this point but it doesn't match either of the two recent snippets:
        "upgrade_state": {
            "ahb_new_version_number": "",
            "ams_new_version_number": "",
            "consistency_request": false,
            "dis_state": 0,
            "err_code": 0,
            "force_upgrade": false,
            "message": "",
            "module": "null",
            "new_version_state": 2,
            "ota_new_version_number": "",
            "progress": "0",
            "sequence_id": 0,
            "status": "IDLE"
        },

**Comment by AdrianGarside** [comment]
https://github.com/greghesp/ha-bambulab/releases/tag/v1.6.0-dev2

**Comment by fmeus** [comment]
Will test it as soon as I can convince HA/HACS to install updates....

Had a quick glance through the changes. new_version_state is equal to 2 when there are no updates available, so I think the check should be

`is_on_fn=lambda self: (self.coordinator.get_model().info.firmware_updates) != 2`

 instead of

`is_on_fn=lambda self: self.coordinator.get_model().info.new_version_state == 1`

Also the assets do not seem to include your changes. v1.6.0-dev2 has been created three days ago....

**Comment by AdrianGarside** [comment]
The only data points we have is that 2 is no firmware update needed, 1 is a firmware updated needed. We currently have no idea what any other values mean and if or when they might be present. So ==1 is the more restrictive check. If we learn there are other values and what they mean I will amend the check.

**Comment by AdrianGarside** [comment]
Not sure what’s up with the release. I’ll check it in the morning (1am here).

**Comment by fmeus** [comment]
> The only data points we have is that 2 is no firmware update needed, 1 is a firmware updated needed. We currently have no idea what any other values mean and if or when they might be present. So ==1 is the more restrictive check. If we learn there are other values and what they mean I will amend the check.

Ah boy, seems I am in need of my morning coffee. Totally glanced over the swapped comparison (!= vs ==). Sorry for that.

**Comment by AdrianGarside** [comment]
Kicked a new release, now really heading to be...
https://github.com/greghesp/ha-bambulab/releases/tag/v1.6.0-dev3

Assets are listed as 3 hours old on this one so hopefully it's good. I confirmed that the -dev2 drop did not have this fix. And this -dev3 drop does. Very odd.

**Comment by fmeus** [comment]
It appears to be working, now if only Bambu Lab would push another firmware update to check the other state.

I did notice errors in HA logfile

     TypeError:` 'str' object cannot be interpreted as an integer

But these appeared tp be related to the start time fix, but it is (based on de code in dev3) missing a conversion

    if data.get("gcode_start_time") is not None:
         self.start_time = get_start_time(     int(      data.get("gcode_start_time")    )     )

**Comment by AdrianGarside** [comment]
Thanks. I can't easily test that code on my P1P. Lost that conversion as I reworked that code to try and get it to work reliably.

**Comment by AdrianGarside** [comment]
> Not sure what’s up with the release. I’ll check it in the morning (1am here).

Human error. I didn't set the branch to develop instead of main when I was publishing. I know now what that looks like since I just did it again this morning :)

**Comment by AdrianGarside** [comment]
Syntax error fixed in https://github.com/greghesp/ha-bambulab/releases/tag/v1.6.0-dev6

**Comment by fmeus** [comment]
Thanks, the log is clean again

**Comment by AdrianGarside** [comment]
Fixed in 1.6.0 official release.
